File: src/weserv/manipulators/gamma.lua
local tonumber = tonumber

--- Gamma manipulator
-- @module gamma
local manipulator = {}

--- Resolve gamma amount.
-- @param gam The given gamma.
-- @return The resolved gamma amount.
function manipulator.resolve_gamma(gam)
    local gamma = tonumber(gam)

    -- Gamma may not be nil and needs to be in the range of 1.0 - 3.0
    if gamma ~= nil and gamma >= 1.0 and gamma <= 3.0 then
        return gamma
    end

    -- Default gamma correction of 2.2 (sRGB)
    return 2.2
end

--- Should this manipulator process the image?
-- @param args The URL query arguments.
-- @return Boolean indicating if we should process the image.
function manipulator.should_process(args)
    return args.gam ~= nil
end

--- Perform gamma image manipulation.
-- @param image The source image.
-- @param args The URL query arguments.
-- @return The manipulated image.
function manipulator.process(image, args)
    local gamma = manipulator.resolve_gamma(args.gam)

    -- Edit the gamma
    if args.has_alpha then
        if not args.is_premultiplied then
            -- Premultiply image alpha channel before gamma transformation
            image = image:premultiply()
            args.is_premultiplied = true
        end

        -- Separate alpha channel
        local image_without_alpha = image:extract_band(0, { n = image:bands() - 1 })
        local alpha = image:extract_band(image:bands() - 1, { n = 1 })
        image = image_without_alpha:gamma({ exponent = 1.0 / gamma }) .. alpha
    else
        image = image:gamma({ exponent = 1.0 / gamma })
    end

    return image
end

return manipulator

File: src/api/processors/gamma.cpp
#include "gamma.h"

namespace weserv::api::processors {

VImage Gamma::process(const VImage &image) const {
    auto gamma = query_->get<float>("gam", 0.0F);

    // Should we process the image?
    if (gamma == 0.0F) {
        return image;
    }

    // Gamma needs to be in the range of 1.0 - 3.0
    if (gamma < 1.0 || gamma > 3.0) {
        // Set gamma to the default correction (sRGB)
        gamma = 2.2;
    }

    // Edit the gamma
    if (image.has_alpha()) {
        // Separate alpha channel
        auto image_without_alpha = image.extract_band(
            0, VImage::option()->set("n", image.bands() - 1));
        auto alpha = image[image.bands() - 1];
        return image_without_alpha
            .gamma(VImage::option()->set("exponent", 1.0 / gamma))
            .bandjoin(alpha);
    }

    return image.gamma(VImage::option()->set("exponent", 1.0 / gamma));
}

}  // namespace weserv::api::processors
